test/Dialect/FIRRTL/dedup.mlir

```diff
@@ -37,6 +37,36 @@ firrtl.circuit "Simple" {
   }
 }
 
+// CHECK-LABEL: "PrivateModulesDifferentConventionsDedup"
+firrtl.circuit "PrivateModulesDifferentConventionsDedup" {
+  firrtl.module private @A() attributes {convention = #firrtl<convention scalarized>} { }
+  firrtl.module private @B() attributes { } { }
+  firrtl.module @PrivateModulesDifferentConventionsDedup() {
+    // CHECK: firrtl.instance a @A()
+    // CHECK: firrtl.instance b @A()
+    firrtl.instance a @A()
+    firrtl.instance b @B()
+  }
+}
+
+
+firrtl.circuit "PublicAndPrivateModulesDedup" {
+  // CHECK-NOT: @A
+  firrtl.module private @A() { }
+  // CHECK:     @B()
+  firrtl.module @B() { }
+  // CHECK-NOT: @C
+  firrtl.module private @C() { }
+  firrtl.module @PublicAndPrivateModulesDedup() {
+    // CHECK: firrtl.instance a @B()
+    // CHECK: firrtl.instance b @B()
+    // CHECK: firrtl.instance c @B()
+    firrtl.instance a @A()
+    firrtl.instance b @B()
+    firrtl.instance c @C()
+  }
+}
+
```

Comment:
Can the PR include tests of these three annotation-based dedup features and how this works with public/private:

1. `firrtl.transforms.MustDedupAnnotation`
2. `firrtl.transforms.NoDedupAnnotation`
3. `firrtl.transforms.DedupGroupAnnotation`

It's possible that the tests could be elided if there is sufficient coverage or the code paths don't warrant it.

Reply:
I don't think the code paths really warrant more tests, the way they operate stays the same with this new feature. MustDedup is a post-transform check that the given list of modules deduped, it doesn't in any way force modules to dedup.  NoDedup modules are still never hashed.  The DedupGroup attribute is hashed like the rest of the attributes on a module, e.g. portTypes.

Reply:
Fantastic. 💯  Thanks for thinking through this.